add carousels for llm params and how-llm-works

heaversm opened the pull request:
Added a PR for this. Not sure how to include properly, as the compile script doesn’t work, but essentially you would just include each like in the img attached.
<img width="559" alt="Screenshot 2023-08-16 at 11 31 46 AM" src="https://github.com/skyfallsin/MIECO/assets/520541/e1851eb9-ed92-4d41-8f40-0f79cbf310b1">

heaversm committed:
add carousels for llm params and how-llm-works

skyfallsin committed:
adding fix for typo

skyfallsin committed:
patching up markdown to carousel conversion

skyfallsin committed:
fixing ai-basics carousel

pdehaan commented on templates/ai/content/ai-basics/_ai-basics-carousel.html:
```diff
@@ -0,0 +1,27 @@
+{% set slides = [
+  { "alt": "Tokenization Visualization","title": "Tokenization", "description": "The LLM starts by breaking down the input text, or 'prompt', into smaller pieces known as tokens.", "image": "/img/ai/carousels/how-llm-works/tokenization.png"},
+  { "alt": "Embedding Visualization","title": "Embedding", "description": "Next, each token is converted into a numerical representation through a process called embedding", "image": "/img/ai/carousels/how-llm-works/embedding.png"},
+  { "alt": "Self-attention Visualization","title": "Self-attention", "description": "The model then adds additional information to these vectors to account for the order of words in the input. Self attention mechanisms allow the model to create context-aware representations of each word.", "image": "/img/ai/carousels/how-llm-works/self-attention.png"},
+  { "alt": "Decoding Visualization","title": "Decoding", "description": "The LLM Decoder will take the output and use it to predict the next token in the sequence.", "image": "/img/ai/carousels/how-llm-works/decoding.png"},
+  { "alt": "Output Visualization","title": "Output", "description": "Finally, the model generates a response.", "image": "/img/ai/carousels/how-llm-works/output.png"}
+]
+%}
```
This looks the same as https://github.com/skyfallsin/MIECO/blob/4903eda1110f1fb46f453ba378b209a45d8d79da/pages/ai/content/ai-basics/_ai-basics-carousel.html#L1-L8

Not sure if there is an easy way to share the contents between the files to keep them DRYer.
In fact, I think the two files are exactly the same. Interesting.

Also, one day we should talk about the Wonderful World of Eleventy and if it'd make this pipeline a lot easier (since it has native support for Nunjucks+Liquid and a bunch of other nice things baked in).

pdehaan replied:
Also, since I was I was looking at this for something else, Nunjucks [`{% macro %}`](https://mozilla.github.io/nunjucks/templating.html#macro)s might be a nice option. We could define the carousel code in one place, then just [`{% import %}`](https://mozilla.github.io/nunjucks/templating.html#import) the carousel `macro` in the templates where we need carousels (passing in the `slides[]` array as a parameter).

heaversm replied:
@skyfallsin - this is a good point, and makes me realize, I have to manually copy these files over from `templates` to `content` - the build script doesn't manually do this.

pdehaan commented on templates/ai/content/ai-basics/index-content.html:
```diff
@@ -1,4 +1,8 @@
+{% include "ai/content/ai-basics/_ai-basics-carousel.html" %}
+</code>
+
 <h4 id="what-is-ai-in-2023">What is AI in 2023?</h4>
```
Not relevant here, but since the eyes see it (and a11y linting will undoubtedly echo it), it looks like we're jumping from an `<h2>` => `<h4>` here (and presumably elsewhere).
Me thinks these should [eventually] be converted to H3 tags to keep them in sequential descending order.

heaversm replied:
Good call - I changed this (and also the build script, which is looking for h4s. But I think this again points out where the build script is a bit brittle @skyfallsin - we're gonna need more flexibility here to use different level h tags and still be able to generate the needed markup / IDs / etc.

pdehaan replied:
Yeah, generally in the past I've used [11ty](https://www.11ty.dev/), with the [markdown-it-anchor](https://www.npmjs.com/package/markdown-it-anchor) plugin, to manage auto-generated `id`s in my heading tags, but then you'd still have to hardcode the anchors from your side nav, so maybe it's moot.
But I think you might be able to simplify a bunch of the plumbing using 11ty.

pdehaan commented on templates/ai/content/ai-basics/_ai-basics-carousel.html:
```diff
@@ -0,0 +1,27 @@
+{% set slides = [
+  { "alt": "Tokenization Visualization","title": "Tokenization", "description": "The LLM starts by breaking down the input text, or 'prompt', into smaller pieces known as tokens.", "image": "/img/ai/carousels/how-llm-works/tokenization.png"},
+  { "alt": "Embedding Visualization","title": "Embedding", "description": "Next, each token is converted into a numerical representation through a process called embedding", "image": "/img/ai/carousels/how-llm-works/embedding.png"},
```
nit: this `description` doesn't have a trailing period, but others do.

pdehaan commented on templates/ai/content/ai-basics/index.md:
```diff
@@ -1,5 +1,9 @@
 
+<code>
+{% include "ai/content/ai-basics/_ai-basics-carousel.html" %}
+</code>
```
Q: why is the carousel code wrapped in an [inline code element](https://developer.mozilla.org/docs/Web/HTML/Element/code)?

heaversm replied:
@pdehaan  - apparently the build script borks unless we do this. (cc @skyfallsin )

heaversm committed:
no message

heaversm committed:
remove duplicate includes, add carousels where needed

heaversm committed:
Merge branch 'main' into mh/34-add-includes-for-carousels-v2

heaversm committed:
Changes:

- fixed content generation script. Remove glabel marked install from build script (leave it for npm / package.json)
- Switches h4s to h3s.
- "Up Next" to "Coming".
-

heaversm committed:
changes:

add all template content to tailwind config
fix decoding carousel img
turn tailwind carousel into a simple scrollable div
rename llm101 carousel to params-carousel for specificity / allowance of multiple carousels

heaversm committed:
fix max width problem introduced with carousel

heaversm commented:
@skyfallsin  - this PR is ready for merge I think - can you review?

heaversm committed:
remove primary link (just expands nav now)

pdehaan commented on pages/ai/content/ai-basics/_ai-basics-carousel.html:
```diff
@@ -1,27 +1,22 @@
-  <div id="item{{loop.index}}" class="carousel-item card card-compact bg-neutral w-72 p-4">
-      <figure class="card-figure">
+  <div id="item{{loop.index}}" class="card card-compact bg-neutral w-72 shrink-0 grow-0 ml-6 first:ml-0">
+      <figure class="card-figure !mt-0 !mb-0">
```
You might be able to just use `!my-0` instead of specifying `mt-*` and `mb-*` separately.
https://tailwindcss.com/docs/margin#add-vertical-margin

Not sure if you'd need to make this change in 4 different files (Ctrl+F says there are 4 matches for `!mt-`), which might be another nod to using DRYer [`macro`](https://mozilla.github.io/nunjucks/templating.html#macro)s.

heaversm replied:
@pdehaan  - feel free to implement this and PR if you like. Otherwise will revisit macros when I have time.

skyfallsin committed:
Merge branch 'main' into mh/34-add-includes-for-carousels

skyfallsin committed:
example broken up carousel for review @mheavers tomorrow

pdehaan commented on templates/ai/content/ai-basics/_sidebar.html:
```diff
@@ -7,7 +7,8 @@
 <li><a href="/ai/content/ai-basics/index.html#what-exactly-is-an-llm">What exactly is an LLM?</a></li>
-<li><a href="/ai/content/ai-basics/index.html#what-are-the-pros--cons-of-using-an-llm">What are the pros & cons of using an LLM?</a></li>
+
+<li><a href="/ai/content/ai-basics/index.html#what-are-the-pros-cons-of-using-an-llm">What are the pros & cons of using an LLM?</a></li>
```
Q: should this be "-pros--cons-` or `-pros-cons-`?
I think currently in staging this is broken and the element/`id` is `<h4 id="what-are-the-pros--cons-of-using-an-llm">What are the pros &amp; cons of using an LLM?</h4>`

So changing this back from a double `--` to single `-` might possibly re-break the sidebar anchor link.

pdehaan commented on templates/ai/content/ai-basics/diagrams/_decoder.html:
```diff
@@ -0,0 +1,19 @@
+  <div class="whitespace-nowrap overflow-x-scroll flex py-6">
+    {% for slide in slides %}
+    <div id="item{{loop.index}}" class="card card-compact bg-neutral w-72 shrink-0 grow-0 ml-6 first:ml-0">
+        <figure class="card-figure !mt-0 !mb-0">
```
Does this need `!mt-0 !mb-0` or does a single `!my-0` work?
